core-bundle/src/Controller/AbstractFragmentController.php

```diff
@@ -75,47 +83,91 @@ protected function getPageModel(): ?PageModel
-    protected function createTemplate(Model $model, string $templateName): Template
+    protected function createTemplate(Model $model, string|null $fallbackTemplateName = null): FragmentTemplate
     {
-        if (isset($this->options['template'])) {
-            $templateName = $this->options['template'];
+        $templateName = $this->getTemplateName($model, $fallbackTemplateName);
+
+        if ($isLegacyTemplate = $this->isLegacyTemplate($templateName)) {
+            // TODO: enable deprecation once existing fragments have been adjusted
+            // trigger_deprecation('contao/core-bundle', '5.0', 'Creating fragments with legacy templates is deprecated and will not work anymore in Contao 6.');
         }
 
-        $request = $this->container->get('request_stack')->getCurrentRequest();
+        // Allow calling render() without a view
+        $this->view = !$isLegacyTemplate ? "@Contao/$templateName.html.twig" : null;
+
+        $onGetResponse = function (FragmentTemplate $template, Response|null $preBuiltResponse) use ($templateName, $isLegacyTemplate): Response {
+            if ($isLegacyTemplate) {
+                // Render using the legacy framework
+                $legacyTemplate = $this->container->get('contao.framework')->createInstance(FrontendTemplate::class, [$templateName]);
+                $legacyTemplate->setData($template->getData());
+
+                $response = $legacyTemplate->getResponse();
+
+                if (null !== $preBuiltResponse) {
+                    return $preBuiltResponse->setContent($response->getContent());
+                }
+
+                $this->markResponseForInternalCaching($response);
 
-        if ($model->customTpl) {
-            // Use the custom template unless it is a back end request
-            if (null === $request || !$this->container->get('contao.routing.scope_matcher')->isBackendRequest($request)) {
-                $templateName = $model->customTpl;
+                return $response;
             }
-        }
 
-        $templateClass = FragmentTemplate::class;
+            // Directly render with Twig
+            $context = $this->container->get('contao.twig.interop.context_factory')->fromData($template->getData());
 
-        // Current request is the main request (e.g. ESI fragment), so we have to replace
-        // insert tags etc. on the template output
-        if ($request === $this->container->get('request_stack')->getMainRequest()) {
-            $templateClass = FrontendTemplate::class;
-        }
+            return $this->render($template->getName(), $context, $preBuiltResponse);
+        };
+
+        $template = new FragmentTemplate($templateName, $onGetResponse);
 
-        /** @var Template $template */
-        $template = $this->container->get('contao.framework')->createInstance($templateClass, [$templateName]);
-        $template->setData($model->row());
+        if ($isLegacyTemplate) {
+            $template->setData((array) $model->row());
```

Comment:
```suggestion
            $template->setData($model->row());
```

The type cast seems redundant, because `row()` always returns an array, doesn‘t it?

Reply:
It's mainly to facilitate testing. In all our current unit test it would yield`null` instead of `[]`. I'm planning on building a test case for content element / module controllers in #4444. We can then refactor the old tests and get rid of this.

Reply:
> In all our current unit test it would yield`null` instead of `[]`.

Then we should fix the tests, shouldn‘t we?

Reply:
IMO, the array cast is good practise as the row method has no native return type. Adding an `array` return type to row method would also be fine with me as an alternative.

Reply:
I agree. Still it seems bad practice to yield `null` in the unit tests even though we know that it is an invalid return value. If we added `: array` as return type, the tests would fail immediately, therefore they should yield the correct return value from the beginning IMHO.

Reply:
> If we added `: array` as return type, the tests would fail immediately

I think the Mock implementation would return an empty array by default then (because it knows the return type of the method).

Reply:
And that is all I‘m asking to be changed in this PR. 🙈

Reply:
Do you want to adjust these things in #4536 as well or do we merge your PR first?

Reply:
see f5bf5458ff5c3f4efd89b9d1c23088a85bba0e62